### PR TITLE
Issue/9560 offline featured image

### LINKED ISSUE #9560: Offline Support: Can't Set a Featured Image while offline.
## Description:

Post Settings -> Set Featured Image -> Pick an image -> Click on Done -> a toast “There is no network available” is shown

## Matching iOS issue:

https://github.com/wordpress-mobile/WordPress-iOS/issues/11433

### PR DESCRIPTION
**Pull request by malinajirka**
Fixes #9560

Note: 
~**- This PR is dependant on https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/1289 which needs to be merged first**~

Adds support for uploading post featured images asynchronously - without blocking the UI.

To test:
Initial steps for all test cases
1. Create a draft
2. Click on the overflow menu in the editor -> Post settings
3. Set a featured image from a local storage

--------------------

Successful upload
4. Notice the local image and a progress bar is shown during the upload
5. Notice the progress bar and it's overlay disappears when the upload finishes

--------------------

Failed upload with retry
4. Notice the local image and a progress bar is shown during the upload
5. Turn on airplane mode before the upload finishes
6. Notice the "tap for options" message
7. Turn off airplane mode
8. Click on the image and choose "Retry uploading"
9. Notice the local image and a progress bar is shown during the upload
10. Notice the progress bar and it's overlay disappears when the upload finishes

--------------------

Failed upload with remove
4. Notice the local image and a progress bar is shown during the upload
5. Turn on airplane mode before the upload finishes
6. Notice the "tap for options" message
7. Click on the image and choose "Remove"
8. Notice the image is removed, upload canceled and the error system notification dismissed

--------------------

Cancel ongoing upload
4. Notice the local image and a progress bar is shown during the upload
5. Click on the image and choose remove
6. Notice the image is removed, upload canceled and the error system notification dismissed

--------------------

Choose another image during ongoing upload
4. Notice the local image and a progress bar is shown during the upload
5. Click on the image and choose featured image option
6. Notice the previous image is removed, upload canceled and the error system notification dismissed
7. The newly picked image is being uploaded

--------------------

Post List label
4. Notice the local image and a progress bar is shown during the upload
5. Go back to the post list and notice "Uploading media" label is shown

--------------------

Remove all media dialog (test this for both Gutenberg and Aztec)
4. Notice the local image and a progress bar is shown during the upload
5. Turn on airplane mode
6. Go back to the editor and click on Update/Save/Publish
7. Publish confirmation dialog is shown -> click on Publish now
8. Notice another dialog is shown -> Click on "Remove all failed media"
9. The failed upload is removed and the error notification dismissed
10. Click on Update/Save/Publish again and notice the post gets published
(Note: It might be worth showing a snackbar message when we remove all the media - the flow seems a bit broken now, since nothing visible happens when you click on "Remove all failed media")

Update release notes:

- [ x ] If there are user facing changes, I have added an item to `RELEASE-NOTES.txt`.

Known issues:
~- If you cancel the upload when the "error uploading media" notification is shown, the notification doesn't get canceled.~
- We don't track any events - https://github.com/wordpress-mobile/WordPress-Android/issues/10151

Before            |  After
:-------------------------:|:-------------------------:
featured-image |  fixed-featured-image

### CODE REVIEW

**Comment by malinajirka** [comment]
@osullivanchris I've added a support for adding a featured image to a post in offline (see before/after gifs). When the user picks a new image, the image is queued and uploaded later.

When the upload fails, the current solution displays a system notification, but keeps showing a circular progress bar in place of the featured image. When the user clicks on the progress bar, they can cancel the upload. I didn't want to replace the circular progress bar with "retry/cancel" buttons or similar as it might look like the user can't leave the screen. However, the current solution doesn't feel great either. Any suggestions?

cancel-featured-image

**Comment by maxme** [comment]
> However, the current solution doesn't feel great either.

I agree, the infinite progress bar feels wrong. 

Could we show the local image instead of the progress bar? Since it's supposed to "just work" when the draft is uploaded. We could replace the "Cancel upload" popover menu by the other "Choose / Remove" menu.

**Comment by maxme** [comment]
Something wrong just happened when I was testing with an existing draft: 

- Airplane mode ON
- Open an existing draft
- Click on the overflow menu in the editor -> Post settings
- Set a featured image from a local storage
- Notice a progress bar is shown in place of the featured image
- Go back to Post List and notice the post upload error
- Edit that draft again
- Turn OFF airplane mode
- Click on the overflow menu in the editor -> Post settings
- Notice a progress bar is shown in place of the featured image
- Back to the editor
- Click on Update
- Wait until the draft is uploaded
- Notice the featured image is not shown in the post list
- Edit that draft again
- Click on the overflow menu in the editor -> Post settings
- Notice the button "Set featured image" 💥 

[demo video](https://bia.is/s/QjTx/offline-featured-image-bug1.webm)

**Comment by osullivanchris** [comment]
Hey @malinajirka I know this issue is in review so I don't want to hold it up too much. However I think there are a few clarifications needed.

1. In the existing version, and while online, there is a blocking loading animation over the gallery (choose photo) view. In the proposed version, the loading takes place on the 'Post settings' screen. I think this is an improvement. I just want to check that you're changing that for both online and offline?
2. Doing it on the post settings screen is good because it doesn't block the user, they can do some other things while the upload takes place. My question is whether the user can leave the 'Post settings' view while uploading is taking place? If the user can leave the post settings screen, we should have an uploading and failed state for the post card on the 'posts' screen like we do for other uploads.

Here is a quick mock for something close to what you currently have. I am using the same linear indeterminate indicator that we have on the posts screen, rather than the circular spinner. I don't think tapping the spinner to bring up a cancel option is intuitive. In this mock, the option is visible on the screen.
upload media fail

Happy to work on this further but showing you as quick as possible to get your input and make sure I understand all aspects.

**Comment by iamthomasbishop** [comment]
At first glance, I don't dislike your approach here, @osullivanchris. I do however think we might want to consider using an existing pattern – in this case, perhaps replicate what exists during upload/error states on the editor canvas itself.

Essentially, that means show the progress indicator docked to the top of the upload in-progress, and if there is an error, show the messaging/actions in an overlay. Example:

IMG_0490

**Comment by malinajirka** [comment]
Thank you so much for the quick input @osullivanchris !!

>In the existing version, and while online, there is a blocking loading animation over the gallery (choose photo) view. In the proposed version, the loading takes place on the 'Post settings' screen. I think this is an improvement. I just want to check that you're changing that for both online and offline?

yes, I moved the upload to the background for both online and offline.

>Doing it on the post settings screen is good because it doesn't block the user, they can do some other things while the upload takes place. My question is whether the user can leave the 'Post settings' view while uploading is taking place? If the user can leave the post settings screen, we should have an uploading and failed state for the post card on the 'posts' screen like we do for other uploads.

Yes, the user can leave the screen. The error/retry state on the posts screen is without a change - the app doesn't differentiate between an image in the post content and a featured image (if either fails, the "error uploading media + retry" is shown). You can see the post list screen on the "after" gif in the description of the PR.

>Here is a quick mock for something close to what you currently have. I am using the same linear indeterminate indicator that we have on the posts screen, rather than the circular spinner. I don't think tapping the spinner to bring up a cancel option is intuitive. In this mock, the option is visible on the screen.

Awesome, thanks! The only thing I'm a bit concerned about is that in offline the user will see the "Upload failed" right away -> I'm not sure they'll realize they can freely leave the screen and continue writing the post content. Wdyt?

**Comment by osullivanchris** [comment]
@malinajirka thanks for the clarifications that all makes sense. 

> Awesome, thanks! The only thing I'm a bit concerned about is that in offline the user will see the "Upload failed" right away -> I'm not sure they'll realize they can freely leave the screen and continue writing the post content. Wdyt?
I'm not too worried about that. We aren't blocking the UI with an overlay or anything. And if the user is has just done a task of trying to upload an image and it failed, they would want to know about it. Is there anything there I'm missing?

I chatted to @iamthomasbishop to get a review and you'll see his sketch above too. Thanks for bringing this pattern to my attention Thomas! 

@malinajirka I think both my mock and Thomas' sketch have more or less the same functionality with different layout. We show the image the user is attempting to upload, a loading progress bar, and the ability to retry or remove if the upload failed. Does this seem do-able from your side, and is there anything you spot with either approach that would be an issue here?

If you're happy to build this, I can do a final high fidelity mock of Thomas' sketch to clarify

**Comment by malinajirka** [comment]
Thank you both! 

>I'm not too worried about that. We aren't blocking the UI with an overlay or anything. And if the user is has just done a task of trying to upload an image and it failed, they would want to know about it.

👍 

>Does this seem do-able from your side, and is there anything you spot with either approach that would be an issue here?
>If you're happy to build this, I can do a final high fidelity mock of Thomas' sketch to clarify

Sounds good, thanks!

**Comment by osullivanchris** [comment]
Having another crack at this. Let me know what you think @malinajirka @iamthomasbishop 

In both the below I have the same actions and states, just different layout. Its tricky because we have this kind of successful or expected-failed state. Where we grab the image, but its not uploaded yet, only queued. I wanted to make this state distinct from something breaking. State 1 is in-progress. State 2 is successfully queued while offline (the aim of this ticket). State 3 is if something breaks or any other generic fail state, and allows the user to retry.

For the mock first I tried what Thomas said. Consistency is a good thing. But I feel like the layout is too rigid of this kind of button lock up
- The layout breaks with truncation and we can't put longer strings in
- actions above images is hard to get perfect because the image can be different heights
- actions above the image may lead to the text being illegible in some cases 
I have mocked it up all the same
option1

Then I tried to make my original approach work, while using patterns we have elsewhere in the app as much as possible (if not specific to media upload flows)
option2

I also had a look at what corresponding information we could display on the post cards for each of these states
card states

**Comment by malinajirka** [comment]
Thank you so much for looking into this! I really like the direction this is going;).

>Its tricky because we have this kind of successful or expected-failed state. Where we grab the image, but its not uploaded yet, only queued. I wanted to make this state distinct from something breaking.

We currently don't have this state in the app. We always try to upload the post (even in offline). We could theoretically check if the device is online and start the upload when it is online and only enqueue the post when it's offline, but I'm not sure it's worth the effort since we probably wouldn't use it in any other part of the app. 
Moreover, we also had a discussion if it makes sense not to even try to perform an action in offline just because the OS told us it doesn't have an internet connection (I think the internet availability check is quite reliable on Android, but AFAIR it's not reliable on iOS. So we'd be never able to duplicate the same behavior/UX on iOS.).

>Then I tried to make my original approach work, while using patterns we have elsewhere in the app as much as possible (if not specific to media upload flows)

Just a couple of thoughts and implementation limitations.

- I'm a bit concerned the actions (retry, delete) might be too small -> If I recall correctly the recommended dimensions for a clickable area is 48dp x 48dp. Wdyt?
- The way the actions line-break in "State 3 alt layout" isn't easy to implement. It's not impossible, but not natively supported. We can line-break the text before the actions or line-break the text of the actions, but not move both the actions to the next line when they don't fit on the previous line.
line-break


>I also had a look at what corresponding information we could display on the post cards for each of these states

Looks good ;), but I'd suggest implementing this in a low priority ticket. I think the current labels should do the job for now and we can improve it later - "Local Changes", "Uploading media - for in progress sate" and "Unable to upload this post's media - for error state". Wdyt?

**Comment by osullivanchris** [comment]
> We currently don't have this state in the app.

Ok consistent with other issues I had a picture of different behaviours and outcomes depending if the user is online/offline. But we don't really start from that point. We just try to upload, and now if the upload fails we want to have a graceful fallback telling the user what has happened and allowing them to retry when appropriate. Let me know if I got anything wrong there.

> I'm a bit concerned the actions (retry, delete) might be too small -> If I recall correctly the recommended dimensions for a clickable area is 48dp x 48dp. Wdyt?

Sure, I'd agree there! But I lifted those from the post cards, so we're already doing it. I'd also expect the clickable area to include the text, so it would be a larger target horizontally at least. I imagine on the post cards we have a tap-able area that's bigger than the icon visually. I hope so anyway? :) 

> The way the actions line-break in "State 3 alt layout" isn't easy to implement. It's not impossible, but not natively supported. We can line-break the text before the actions or line-break the text of the actions, but not move both the actions to the next line when they don't fit on the previous line.

Ok I'll keep that in mind, thanks!

> Looks good ;), but I'd suggest implementing this in a low priority ticket.

Yeah lets leave the card states out of this issue for now. We have other ongoing discussions about that anyway.

I will revert shortly with another mock to discuss!

**Comment by osullivanchris** [comment]
Dug around the existing app again and found a pattern that suits what we need for this. Don't know why I didn't encounter it before, my previous version of Thomas' sketch was based on the upload media UI. But the image upload UI within the editor looks closer to what he was sketching and gets us a good enough solution

I've kept the error message exactly the same as a normal image so you can lift that code hopefully. For the loading state, the current one I think has a white overlay and no text. I changed the overlay to dark and added text
(a) to explain what's happening
(b) to give the option to cancel uploading this image per the build that you showed
Featured image proposed

For both overlays we then need a menu to appear if the user taps. Here are the two menus. 

- On the right, the 'retry uploading' menu is exactly what appears with an existing image error in the editor.
- On the left, the menu I have used is the one you get if you already have a featured image

Featured image options

I think I've been spending too long trying to find the 'perfect' logical solution to these issues. There are some minor issues with this solution. However I think its at the good enough level to unblock this particular issue. It reuses existing code which hopefully helps you build this quicker, and it achieves the desired user outcomes
- tells the user the image is loading
- the user can continue working elsewhere
- user can tap to cancel uploading
- communicates the error state with options to resolve.

If we want to improve further (icons, text styles, copy in the buttons) we can open a separate issue.

If this gets your thumbsup I'll consider this done from my side @malinajirka

**Comment by loremattei** [comment]
Hey! I'm moving this to 12.9 since 12.8 has been cut. If you want this to make it into 12.8, please feel free to move it back, target the release/12.8 branch and ping me to build a new beta.

**Comment by malinajirka** [comment]
@maxme It's ready for another round, thank you 🙇.

I've done a lot of changes. [The issue](https://github.com/wordpress-mobile/WordPress-Android/pull/10079#issuecomment-504911161) you've mentioned got fixed via another PR. I also added a check so we show the "Retry all/Remove all failed media" dialog in this scenario to keep consistency with the editor.

The PR contains too many changes :(. Let me know if reviewing it is to difficult and I'll try to cherry-pick some of the changes and divide it into more PRs. (not sure if it's worth it and if it's even feasible)

File: RELEASE-NOTES.txt [hunk 1]
```diff
@@ -3,6 +3,8 @@
  
 12.8
 -----
+* Setting featured image on a post doesn't require network connection
+Add percentage bar to following Stats cards - Posts and Pages, Authors, Tags and Categories
```
Comment (maxme) [inline RELEASE-NOTES.txt]: Looks like there was a merge problem here.

**Comment by maxme** [comment]
> The issue you've mentioned got fixed via another PR.

Was it fixed via another PR and is also fixed in this branch? I'm asking this because I found a similar issue:

- In airplane mode, create a new post.
- Add a featured image.
- Tap publish.
- Wait for the error (not sure why the post is uploading for so long since the device is in airplaine mode?)
- Turn airplane mode off, and tap "retry" on the failed upload.
- Wait for the post to be published.
- Open the published post and notice the featured image disappeared.

[Demo video](https://bia.is/s/VUrp/disappearing-featured-image.webm)

**Comment by malinajirka** [comment]
@maxme I've just tested this several times on a Pixel 2 emulator (Android 9) and Pixel 3 hw device and it worked as expected. Do you have any more info which might be relevant to the issue?

P.S. What I encountered one time was that the upload never failed/timed out so I had to restart the app. But I don't think it's relevant to this issue and I'm not able to reproduce it.

File: RELEASE-NOTES.txt (hunk 1 above)
Comment (malinajirka) [reply RELEASE-NOTES.txt]: Fixed in https://github.com/wordpress-mobile/WordPress-Android/pull/10079/commits/709fd35f7401e55810e1e7c0ee944356f8c748f4

**Comment by malinajirka** [comment]
As we discussed on Slack I'm still not able to reproduce this issue without a debugger. However, when I put a breakpoint [at the right spot](https://github.com/wordpress-mobile/WordPress-Android/blob/394d379fa0a9d8f371601a367effcde5d7cb1f70/WordPress/src/main/java/org/wordpress/android/ui/uploads/UploadService.java#L773), I can reproduce it sometimes.

I think the issue is that we sometimes start the UploadService with "Retry upload" intent more than once. The first invocation will 
1. [Get failed media uploads from the database](https://github.com/wordpress-mobile/WordPress-Android/blob/3ffcb23f7ea133357b5a8127e616bc1d4f0af29d/WordPress/src/main/java/org/wordpress/android/ui/uploads/UploadService.java#L780)
2. [Set their upload state to QUEUED](https://github.com/wordpress-mobile/WordPress-Android/blob/3ffcb23f7ea133357b5a8127e616bc1d4f0af29d/WordPress/src/main/java/org/wordpress/android/ui/uploads/UploadService.java#L785)

However, when the second invocation tries to get the failed media uploads from the database, the database returns an empty set as all the media are already in QUEUED state (not FAILED). Therefore, the post is [enqueued for upload](https://github.com/wordpress-mobile/WordPress-Android/blob/3ffcb23f7ea133357b5a8127e616bc1d4f0af29d/WordPress/src/main/java/org/wordpress/android/ui/uploads/UploadService.java#L812) ignoring its media.
I've added [this simple check](https://github.com/wordpress-mobile/WordPress-Android/blob/3ffcb23f7ea133357b5a8127e616bc1d4f0af29d/WordPress/src/main/java/org/wordpress/android/ui/uploads/UploadService.java#L767) to prevent this issue from happening.

@maxme It's ready for another round. The issue wasn't related to the changes made in this PR, however, it made the bug easier to reproduce. It should be fixed in https://github.com/wordpress-mobile/WordPress-Android/pull/10079/commits/3ffcb23f7ea133357b5a8127e616bc1d4f0af29d. This fix should also fix the following issue -> https://github.com/wordpress-mobile/WordPress-Android/issues/9704.

Thank you for helping me reproduce the issue and find the main cause 🙇

**Comment by malinajirka** [comment]
We found another scenario under which the post was uploaded and the featured image got removed.
Steps
1. Turn on airplane mode
2. Create a draft and add a featured image
3. Create a second draft and [the same](url) featured image
4. Go back to Post list
5. Turn off airplane mode
6. Notice the second draft gets uploaded without its featured image

This issue should be fixed in https://github.com/wordpress-mobile/WordPress-Android/pull/10204

However, there is one more issue we'll need to fix when 10204 is merged into this branch.
1. Turn on airplane mode
2. Create a draft 
3. add an image into the content of the post
3. Set the same image as featured image
4. Go back to Post list
5. Turn off airplane mode
6. Notice the draft gets uploaded without its featured image

File: WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java [hunk 2]
```diff
@@ -3095,9 +3108,15 @@ public void onActivityResult(int requestCode, int resultCode, Intent data) {
                 case RequestCodes.PHOTO_PICKER:
                 case RequestCodes.STOCK_MEDIA_PICKER_SINGLE_SELECT:
                     // user chose a featured image
-                    if (resultCode == RESULT_OK && data.hasExtra(PhotoPickerActivity.EXTRA_MEDIA_ID)) {
-                        long mediaId = data.getLongExtra(PhotoPickerActivity.EXTRA_MEDIA_ID, 0);
-                        setFeaturedImageId(mediaId);
+                    if (resultCode == RESULT_OK) {
```
[the remark this answers is not in the document]
Comment (malinajirka) [reply WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java]: Fixed in https://github.com/wordpress-mobile/WordPress-Android/pull/10079/commits/5f63df8e1faa2a3f806ac45e95649e616f323be8

**Comment by malinajirka** [comment]
>The feature works as expected (if we ignore the issues with media uploads).

The issue with media uploads should be fixed now. It'd be great if you tested it as I can't reproduce it without a debugger. 🤞

**Pull request merged**